### PR TITLE
lib: location: Add LTE system mode sanity check

### PR DESCRIPTION
**Pull request by glarsennordic**
method_gnss now checks that the currently selected
modem system mode supports GNSS when performing
location requests. If it does not, the location
request will print an error and fail.

method_gnss will also now check that the currently
selected modem system mode supports LTE when
attempting to download GNSS assistance data.
If not, the download attempt will print an error
and fail.

IRIS4187

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

### CODE REVIEW

**Comment by glarsennordic** [comment]
DNM until I get a chance to test this with a few more samples

**Review by plskeggs — APPROVED** [review]
LGTM

File: doc/nrf/releases/release-notes-changelog.rst [hunk 1]
```diff
@@ -448,6 +448,11 @@ Other libraries
     The library is no longer directly referenced from the Application Event Manager.
     Instead, it uses the Application Event Manager hooks to connect with the manager.
 
+* :ref:`lib_location`:
+
+  * Added sanity check to double check that the currently selected LTE system mode supports GNSS when using the GNSS location method for a location request.
```
Comment (peknis) [inline doc/nrf/releases/release-notes-changelog.rst:453]:
```suggestion
  * Added sanity check to verify that the currently selected LTE system mode supports GNSS when using the GNSS location method for a location request.
```

File: doc/nrf/releases/release-notes-changelog.rst [hunk 2]
```diff
@@ -448,6 +448,11 @@ Other libraries
...
+    An error will be printed and the location request will return an error code if the sanity check fails.
```
Comment (peknis) [inline doc/nrf/releases/release-notes-changelog.rst:454]:
```suggestion
    An error is printed and the location request returns an error code if the sanity check fails.
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: doc/nrf/releases/release-notes-changelog.rst (hunk 1 above)
Comment (hiltunent) [reply doc/nrf/releases/release-notes-changelog.rst:453]: Nitpicking: I would replace "LTE" with "modem" because LTE spec does certainly not say anything about GNSS.

File: lib/location/method_gnss.c [hunk 3]
```diff
@@ -717,6 +717,20 @@ static void method_gnss_positioning_work_fn(struct k_work *work)
...
 	int err;
+	enum lte_lc_system_mode mode;
+
+	err = lte_lc_system_mode_get(&mode, NULL);
+	if (err) {
+		LOG_ERR("Could not read LTE system mode. Error: %d", err);
+		return err;
+	}
+
+	if (mode != LTE_LC_SYSTEM_MODE_GPS &&
```
Comment (hiltunent) [inline lib/location/method_gnss.c:728]: Looks good to me. One could also check that the system mode is not LTE_LC_SYSTEM_MODE_GPS is A-GPS or P-GPS is enabled. But I don't know if that is necessary. In such case I believe A-GPS request would just fail and GNSS position would be returned anyway. So definitely the check you just implemented is more important because there is no way to get a fix without GNSS being enabled.

**Review by hiltunent — APPROVED** [review]

File: lib/location/method_gnss.c (hunk 3 above)
Comment (glarsennordic) [reply lib/location/method_gnss.c:728]: I have just added such checks, and moved the checks to their own static inline functions. Let me know if these changes are acceptable!

**Review by peknis — APPROVED** [review]

File: lib/location/method_gnss.c [hunk 4]
```diff
@@ -109,6 +109,54 @@ static void method_gnss_pgps_ext_work_fn(struct k_work *item);
...
+
+	return 0;
+}
+
+static inline int check_modem_supports_lte(void)
+{
+	int err;
+	enum lte_lc_system_mode mode;
+
+	err = lte_lc_system_mode_get(&mode, NULL);
```
Comment (jhirsi) [inline lib/location/method_gnss.c:139]: Was thinking if lte_lc_lte_mode_get() would be better to do the same? It is returning "none" in plain gnss mode.

**Review by jhirsi — APPROVED** [review]

File: lib/location/method_gnss.c [hunk 5]
```diff
@@ -109,6 +109,54 @@ static void method_gnss_pgps_ext_work_fn(struct k_work *item);
 
 static int fixes_remaining;
 
+static inline int check_modem_supports_gnss(void)
```
Comment (hiltunent) [inline lib/location/method_gnss.c:112]: I don't remember if it possible for neither MQTT nor REST be defined. If so, compiler warning about unused function might pop up? Maybe consider putting this function definition inside ifdef MQTT || MQTT.

**Review by hiltunent — APPROVED** [review]

File: lib/location/method_gnss.c (hunk 4 above)
Comment (glarsennordic) [reply lib/location/method_gnss.c:139]: Good point. Sorry for the delay on the response, I had another project come up and take precedence.

I'll give this a shot and test thoroughly

File: lib/location/method_gnss.c (hunk 5 above)
Comment (glarsennordic) [reply lib/location/method_gnss.c:112]: `check_modem_supports_gnss` is always referenced by `method_gnss_location_get`, so I don't think that will be an issue, but `check_modem_supports_lte` is conditionally referenced. However even with `MQTT || REST` guaranteed, I believe it is possible for the function to not be referenced. I also worry about the potential upkeep implications down the line.

Fortunately, it seems that my use of the inline keyword solves this issue; No warnings are generated when I create a static inline function `check_modem_supports_funny_business` with no references to it. Not sure what we would do if inlining were not an option, but it seems workable in this case.

Does just moving forward with the functions inlined seem reasonable to you, @hiltunent ?

Comment (tokangas) [reply lib/location/method_gnss.c:112]: `check_modem_supports_gnss()` is only called from a single place, so there making the function inline doesn't probably make any difference. But `check_modem_supports_lte()` on the other hand is called from three places and having the function inline unnecessarily increases the flash consumption, because the same code will be on flash three times. Because of that flagging would be preferable.

File: lib/location/method_gnss.c [hunk 6]
```diff
@@ -109,6 +109,48 @@ static void method_gnss_pgps_ext_work_fn(struct k_work *item);
...
+	}
+
+	if (mode == LTE_LC_LTE_MODE_NONE) {
+		LOG_ERR("LTE is disabled. GNSS assistance data downloads are not possible.");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
```
Comment (tokangas) [inline lib/location/method_gnss.c:152]: Extra empty line.

File: lib/location/method_gnss.c [hunk 7]
```diff
@@ -202,8 +244,15 @@ void method_gnss_lte_ind_handler(const struct lte_lc_evt *const evt)
 #if defined(CONFIG_NRF_CLOUD_MQTT)
 static void method_gnss_agps_request_work_fn(struct k_work *item)
 {
-	int err = nrf_cloud_agps_request(&agps_request);
+	int err;
 
+	err = check_modem_supports_lte();
+	if (err) {
+		LOG_ERR("nRF Cloud A-GPS request failed");
```
Comment (tokangas) [inline lib/location/method_gnss.c:251]: You could add the reason (LTE disabled/not supported) in the error message.

File: lib/location/method_gnss.c [hunk 8]
```diff
@@ -226,6 +275,12 @@ static void method_gnss_agps_request_work_fn(struct k_work *item)
 		.fragment_size = 0
 	};
 
+	err = check_modem_supports_lte();
+	if (err) {
+		LOG_ERR("nRF Cloud A-GPS request failed");
```
Comment (tokangas) [inline lib/location/method_gnss.c:280]: You could add the reason (LTE disabled/not supported) in the error message.

File: lib/location/method_gnss.c [hunk 9]
```diff
@@ -305,6 +360,12 @@ static void method_gnss_pgps_request_work_fn(struct k_work *item)
 		.fragment_size = 0
 	};
 
+	err = check_modem_supports_lte();
+	if (err) {
+		LOG_ERR("nRF Cloud P-GPS request failed");
```
Comment (tokangas) [inline lib/location/method_gnss.c:365]: You could add the reason (LTE disabled/not supported) in the error message.

File: lib/location/method_gnss.c [hunk 10]
```diff
@@ -739,6 +800,11 @@ int method_gnss_location_get(const struct location_method_config *config)
 {
 	int err;
 
+	err = check_modem_supports_gnss();
+	if (err) {
+		return err;
```
Comment (tokangas) [inline lib/location/method_gnss.c:805]: Maybe it would be good to have a warning or error trace here as well. If this check fails, the current system mode doesn't match the Location library method configuration, so one of them is incorrect.

File: lib/location/method_gnss.c (hunk 5 above)
Comment (trantanen) [reply lib/location/method_gnss.c:112]: While having `check_modem_supports_gnss()` as inline is technically fine, I'm not sure if we tend to have this type of functions with few if-statements as inline normally. If somebody starts using it in another place, is (s)he going to remove the inline part then? I guess that might easily be missed.

Comment (glarsennordic) [reply lib/location/method_gnss.c:112]: I am not 100% certain flagging is preferable. I agree that increasing the flash usage is non-ideal. I also agree that relying on inlining to prevent compiler warnings seems wrong.

However, I don't think flagging is a good solution either. The exact inclusion conditions for `check_modem_supports_lte` is `(AGPS && !AGPS_EXTERNAL && (MQTT || REST)) || (PGPS && !MQTT)` due to the various conditions under which the functions which rely upon it may be enabled. I worry that this logic will be difficult for a reader to follow and that it will be difficult to ensure the logic is updated if the inclusion conditions ever change.

After some thought, the best solution, I think would be to move the functions to `location_utils.c` and give them header definitions, leaving them non-inline.

I will switch to this approach.

File: lib/location/method_gnss.c (hunk 7 above)
Comment (glarsennordic) [reply lib/location/method_gnss.c:251]: I will do this for the call to `check_modem_supports_gnss` as well

**Review by hiltunent — APPROVED** [review]

**Review by tokangas — APPROVED** [review]

File: lib/location/method_gnss.c (hunk 5 above)
Comment (glarsennordic) [reply lib/location/method_gnss.c]: @trantanen I think I have resolved the issue, do you mind taking a look at my latest approach and checking whether it makes sense to you?

File: lib/location/method_gnss.c [hunk 11]
```diff
@@ -226,6 +235,14 @@ static void method_gnss_agps_request_work_fn(struct k_work *item)
 		.fragment_size = 0
 	};
 
+	err = location_utils_check_modem_supports_lte();
+	if (err) {
+		LOG_ERR("nRF Cloud A-GPS request failed: %s",
+			(err == -EINVAL) ? "Current modem system mode does not support LTE"    :
+					   "Current modem system mode could not be determined");
```
Comment (trantanen) [inline lib/location/method_gnss.c:242]: This is nitpicking but instead of copying these strings into few places, I'd think of logging this in location utils functions. There is already a trace for `-EBUSY` case but not for `-EINVAL`. If this is left as it is, I hope compiler optimizes these strings as possible.

File: lib/location/method_gnss.c (hunk 5 above)
Comment (trantanen) [reply lib/location/method_gnss.c]: Yep, this is fine for me.

File: lib/location/method_gnss.c (hunk 11 above)
Comment (trantanen) [reply lib/location/method_gnss.c:242]: > And, in retrospect, I tend to agree with @tokangas on this one, it makes the functions in `location_utils` cleaner (their sole purpose is to detect modem feature support, rather than serve dual purpose by detecting and alerting to problems. 

Ideally yes. However, in practice the most accurate information is often within the function that first observe that error occurred so writing an error message there kind of makes sense. From design perspective, what would happen if those util functions were changed and there would be 3rd and maybe 4th return value. Or if there would be one or two more places calling these functions.

> Personally, I think the benefits of the more explicit, easier to interpret error message is outweighed by the cost of ~300 bytes of flash.

I don't see why you couldn't have the same error message content in the utils functions in this case. And while 200-300 bytes is quite little, we are using it here a little unnecessarily.

> Perhaps for this same reason, I should remove the error messages on the `-EBUSY` cases?).

If we keep it the way it's implemented now, I think you should do this.

But let's see what @tokangas says. I guess two Tommi's need to have a chat :) But let's not make this too big of a deal as in this specific case, it really doesn't matter that much which way this is done.

Comment (trantanen) [reply lib/location/method_gnss.c:242]: > The util functions have no knowledge of whether they are being called by AGPS or by PGPS functions, so they cannot indicate which of the two failed. We could make a generic message mentioning both, but then if any other part of location library decides to use the util functions, the error message would no longer be accurate. Passing in the calling reason as an argument is another option, but that would make the code hard to follow, IMO

Yep, but I'm assuming you have two logs as you wrote earlier: `"A-GPS request failed" and "Error: current modem system mode does not support LTE"`.

I don't see a big problem with this given it's an error and it can raise attention for someone who is looking at the logs. It's a log anyway and not the real output of an end application.

Comment (trantanen) [reply lib/location/method_gnss.c:242]: Let's leave it as it is. And let's not add another error trace in the utils functions because it may not be an error condition for the caller.

**Review by trantanen — APPROVED** [review]

File: lib/location/method_gnss.c (hunk 11 above)
Comment (tokangas) [reply lib/location/method_gnss.c:242]: I like the current approach. If you think from `location_utils_check_modem_supports_gnss()` point of view, it logs an error if checking for GNSS support fails. If support can be queried from modem, but GNSS is not supported, it's not considered an error. The check was successful and GNSS is not supported.

From `method_gnss_location_get()` point of view, it is considered an error if checking for GNSS support fails or GNSS is not supported. But a different function might also call `location_utils_check_modem_supports_gnss()`, but not consider missing GNSS support an error.

Most likely the compiler optimizes the duplicate strings out, as long as the strings are identical. And these are only traces, so it's also possible to disable them using Kconfig.

**Comment by glarsennordic** [comment]
I will be doing GNSS/walk tests of another PR next week, so I am going to hold off on merging this so that I can throw it in with the other tests

**Review by plskeggs — APPROVED** [review]

**Comment by glarsennordic** [comment]
Will rebase and update tests after https://github.com/nrfconnect/sdk-nrf/pull/8766 merges

File: lib/location/location_utils.h [hunk 12]
```diff
@@ -61,4 +61,22 @@ const char *location_utils_nrf_cloud_jwt_generate(void);
  */
 void location_utils_systime_to_location_datetime(struct location_datetime *datetime);
 
+/**
+ * @brief Check whether the currently active LTE link controller system mode supports GNSS
```
Comment (peknis) [inline lib/location/location_utils.h:65]:
```suggestion
 * @brief Check whether the currently active LTE link controller system mode supports GNSS.
```

File: lib/location/location_utils.h [hunk 13]
```diff
@@ -61,4 +61,22 @@ const char *location_utils_nrf_cloud_jwt_generate(void);
...
+ *
+ * @return 0		If GNSS is supported
+ * @return -EBUSY	If the LTE link controller system mode could not be determined
+ * @return -EINVAL	If GNSS is not supported
+ */
+int location_utils_check_modem_supports_gnss(void);
+
+/**
+ * @brief Check whether the currently active LTE link controller system mode supports LTE
```
Comment (peknis) [inline lib/location/location_utils.h:74]:
```suggestion
 * @brief Check whether the currently active LTE link controller system mode supports LTE.
```

**Review by peknis — CHANGES_REQUESTED** [review]
